src/oge/emissions.py

```diff
@@ -286,48 +296,89 @@ def adjust_fuel_and_emissions_for_CHP(df):
+    Calculates a unique EAF for each subplant-month-fuel.
+
+
+    Args:
+        df (pd.DataFrame): dataframe fuel_consumed_mmbtu,
+        fuel_consumed_mmbtu_for_electricity, and net_generation_mwh columns, mapped to
+        subplant_id
+
+    Returns:
+        pd.DataFrame: the input dataframe with an `electric_allocation_factor` column added
     """
+    groupby_cols = ["plant_id_eia", "subplant_id", "report_date"]
+    # for EIA-923 we want to group on energy_source_code as well, but this does not
+    # exist in CEMS
+    if "energy_source_code" in df.columns:
+        groupby_cols += ["energy_source_code"]
+
+    df_subplant = (
+        df.groupby(groupby_cols)[
+            [
+                "fuel_consumed_mmbtu",
+                "fuel_consumed_for_electricity_mmbtu",
+                "net_generation_mwh",
+            ]
+        ]
+        .sum()
+        .reset_index()
+    )
 
     # calculate the gross thermal output
     # 0.8 is an assumed efficiency factor used by eGRID
-    df["gross_thermal_output_for_heating_mmbtu"] = (
+    df_subplant["gross_thermal_output_for_heating_mmbtu"] = (
         chp_gross_thermal_output_efficiency
-        * (df["fuel_consumed_mmbtu"] - df["fuel_consumed_for_electricity_mmbtu"])
+        * (
+            df_subplant["fuel_consumed_mmbtu"]
+            - df_subplant["fuel_consumed_for_electricity_mmbtu"]
+        )
     )
     # calculate the useful thermal output
     # 0.75 is an assumed efficiency factor used by eGRID
-    df["useful_thermal_output_mmbtu"] = (
+    df_subplant["useful_thermal_output_mmbtu"] = (
         chp_useful_thermal_output_efficiency
-        * df["gross_thermal_output_for_heating_mmbtu"]
+        * df_subplant["gross_thermal_output_for_heating_mmbtu"]
     )
 
     # convert generation to mmbtu
-    df["generation_mmbtu"] = df["net_generation_mwh"] * ConversionFactors.mwh_to_mmbtu
+    df_subplant["generation_mmbtu"] = (
+        df_subplant["net_generation_mwh"] * ConversionFactors.mwh_to_mmbtu
+    )
 
     # calculate the electric allocation factor
-    df["electric_allocation_factor"] = df["generation_mmbtu"] / (
-        df["generation_mmbtu"] + df["useful_thermal_output_mmbtu"]
+    df_subplant["electric_allocation_factor"] = df_subplant["generation_mmbtu"] / (
+        df_subplant["generation_mmbtu"] + df_subplant["useful_thermal_output_mmbtu"]
     )
 
     # if the allocation factor < 0, set to zero
-    df.loc[df["electric_allocation_factor"] < 0, "electric_allocation_factor"] = 0
+    df_subplant.loc[
+        df_subplant["electric_allocation_factor"] < 0, "electric_allocation_factor"
+    ] = 0
     # if the allocation factor > 1, set to one
-    df.loc[df["electric_allocation_factor"] > 1, "electric_allocation_factor"] = 1
+    df_subplant.loc[
+        df_subplant["electric_allocation_factor"] > 1, "electric_allocation_factor"
+    ] = 1
```

Comment:
you can use the [`clip`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.clip.html) method here to assign values outside boundary to boundary values.

Reply:
I didn't know about clip! Just implemented